test/quadiron_c_utest.cpp

```diff
@@ -0,0 +1,261 @@
+    }
+    std::vector<uint8_t*> parity(n_outputs);
+    std::vector<int> wanted_idxs(n_outputs);
+
+    for (int i = 0; i < n_data; i++) {
+        data[i] = static_cast<uint8_t*>(malloc(block_size + metadata_size));
+        ref_data[i] = static_cast<uint8_t*>(malloc(block_size));
+        randomize_buffer(data[i] + metadata_size, block_size);
+        memcpy(ref_data[i], data[i] + metadata_size, block_size);
+    }
+
+    for (int i = 0; i < n_outputs; i++) {
+        parity[i] = static_cast<uint8_t*>(malloc(block_size + metadata_size));
+    }
+
+    // we want all parities
+    for (int i = 0; i < n_outputs; i++) {
+        wanted_idxs[i] = 1;
+    }
+
+    ASSERT_EQ(
+        quadiron_fnt32_encode(
+            inst, data.data(), parity.data(), wanted_idxs.data(), block_size),
+        0);
+
+    copy(
+        missing_idxs.begin(),
+        missing_idxs.end(),
+        std::ostream_iterator<int>(std::cout, " "));
+
+    for (int i = 0; i < n_data; i++) {
+        if (missing_idxs[i]) {
+            memset(data[i], 0, block_size + metadata_size);
+        }
+    }
+
+    for (int i = 0; i < n_parities; i++) {
+        if (missing_idxs[n_data + i]) {
+            memset(parity[i], 0, block_size + metadata_size);
+        }
+    }
+
+    ASSERT_EQ(
+        quadiron_fnt32_decode(
+            inst, data.data(), parity.data(), missing_idxs.data(), block_size),
+        0);
+
+    for (int i = 0; i < n_data; i++) {
+        // quadiron_hex_dump(data[i] + metadata_size, 100);
+        // quadiron_hex_dump(ref_data[i], 100);
+        ASSERT_EQ(memcmp(data[i] + metadata_size, ref_data[i], block_size), 0);
+    }
+
+    std::cout << "reconstruct data ";
+    for (int i = 0; i < n_data; i++) {
+        if (missing_idxs[i]) {
+            std::cout << i << " " << std::flush;
+            ASSERT_EQ(
+                quadiron_fnt32_reconstruct(
+                    inst,
+                    data.data(),
+                    parity.data(),
+                    missing_idxs.data(),
+                    i,
+                    block_size),
+                0);
+        }
+    }
+
+    std::cout << "parity ";
+    for (int i = 0; i < n_parities; i++) {
+        if (missing_idxs[n_data + i]) {
+            std::cout << i << " " << std::flush;
+            ASSERT_EQ(
+                quadiron_fnt32_reconstruct(
+                    inst,
+                    data.data(),
+                    parity.data(),
+                    missing_idxs.data(),
+                    n_data + i,
+                    block_size),
+                0);
+        }
+    }
+
+    std::cout << std::endl;
+
+    for (int i = 0; i < n_data; i++) {
+        free(data[i]);
+        free(ref_data[i]);
+    }
+
+    for (int i = 0; i < n_outputs; i++) {
+        free(parity[i]);
+    }
+
+    quadiron_fnt32_delete(inst);
+}
+
+void test_all_decodable_scenarios(int k, int m, int systematic)
+{
+    for (int i = 0; i <= m; i++) {
+        std::vector<std::vector<int>> combinations =
```

Comment:
use auto: `auto combinations = generate_combinations(k + m, i);`

Could be const maybe.